libtrap/src/ifc_tcpip.c

```diff
@@ -960,419 +958,533 @@ static int client_socket_connect(void *priv, const char *dest_addr, const char *
-      break;
-   case TRAP_HALFWAIT:
-      /*
-       * wait 1s in a loop for select(),
-       * do not change timeout for sem_timedwait in connphase
-       */
-      trap_set_timeouts(1000000, &tv, NULL);
-      break;
-   default:
-      /*
-       * set timeout (can be 0 - nowait or any positive number) for select(),
-       * do not change timeout for sem_timedwait in connphase
-       */
-      trap_set_timeouts(timeout, &tv, NULL);
-      break;
-   }
+again:
+   sent = send(c->sd, c->sending_pointer, c->pending_bytes, MSG_NOSIGNAL);
 
-   /* Check connected clients and wait for them when blocking */
-   result = tcpip_sender_conn_phase(c, ts_p);
-   if (result != TRAP_E_OK) {
-      /* it is useless to send when nobody receives */
-      goto exit;
+   if (sent < 0) {
+      /* Send failed */
+      if (priv->is_terminated != 0) {
+         return TRAP_E_TERMINATED;
+      }
+      switch (errno) {
+      case EBADF:
+      case EPIPE:
+      case EFAULT:
+         return TRAP_E_IO_ERROR;
+      case EAGAIN:
+         goto again;
+      default:
+         VERBOSE(CL_VERBOSE_OFF, "Unhandled error from send in send_data (errno: %i)", errno);
+         return TRAP_E_IO_ERROR;
+      }
+   } else {
+      c->pending_bytes -= sent;
+      c->sending_pointer = (uint8_t *) c->sending_pointer + sent;
+
+      /* Client received whole buffer */
+      if (c->pending_bytes <= 0) {
+         del_index(&buffer->clients_bit_arr, cl_id);
+         if (buffer->clients_bit_arr == 0) {
+            priv->ctx->counter_send_buffer[priv->ifc_idx]++;
+            pthread_cond_broadcast(&priv->cond);
+         }
+
+         /* Assign client the next buffer in sequence */
+         c->assigned_buffer = (c->assigned_buffer + 1) % priv->buffer_count;
+      }
    }
+   return TRAP_E_OK;
+}
 
-   FD_ZERO(&disset);
-   FD_ZERO(&set);
+/**
+ * \brief This function runs in a separate thread. It handles sending data
+          to connected clients for TCPIP and UNIX interfaces.
+ * \param[in] priv pointer to interface private data
+ */
+static void *sending_thread_func(void *priv)
+{
+   uint32_t i, j;
+   int res;
+   int maxsd = -1;
+   fd_set set, disset;
+   client_t *cl;
+   buffer_t *assigned_buffer;
+   uint8_t buffer[DEFAULT_MAX_DATA_LENGTH];
+   uint64_t send_entry_time;
+   uint64_t send_exit_time;
+   uint8_t client_ready;
 
-   /*
-    * add term_pipe for reading into the disconnect client set
-    */
-   FD_SET(c->term_pipe[0], &disset);
-   if (maxsd < c->term_pipe[0]) {
-      maxsd = c->term_pipe[0];
-   }
+   tcpip_sender_private_t *c = (tcpip_sender_private_t *) priv;
 
-   for (i = 0, j = 0; i < c->clients_arr_size; ++i) {
-      cl = &c->clients[i];
-      if (j == c->connected_clients) {
-         break;
+   while (1) {
+      if (c->is_terminated != 0) {
+         pthread_exit(NULL);
       }
-      if (cl->sd <= 0) {
-         /* not connected client */
+      if (c->connected_clients == 0) {
+         usleep(100000);
```

Comment:
This probably should not be hardcoded, use define which can be easily configured.